langchain/src/stores/message/firestore.ts

```diff
@@ -0,0 +1,160 @@
+import * as firebaseAdmin from "firebase-admin";
+import { getApps, initializeApp } from "firebase-admin/app";
+
+import {
+  getFirestore,
+  DocumentData,
+  Firestore,
+  DocumentReference,
+  FieldValue,
+} from "firebase-admin/firestore";
+import {
+  StoredMessage,
+  BaseMessage,
+  BaseListChatMessageHistory,
+} from "../../schema/index.js";
+import {
+  mapChatMessagesToStoredMessages,
+  mapStoredMessagesToChatMessages,
+} from "./utils.js";
+
+export interface FirestoreDBChatMessageHistory {
+  collectionName: string;
+  sessionId: string;
+  userId: string;
+  appIdx?: number;
+  config?: firebaseAdmin.AppOptions;
+}
+export class FirestoreChatMessageHistory extends BaseListChatMessageHistory {
+  lc_namespace = ["langchain", "stores", "message", "firestore"];
+
+  private collectionName: string;
+
+  private sessionId: string;
+
+  private userId: string;
+
+  private appIdx: number;
+
+  private config: firebaseAdmin.AppOptions;
+
+  private firestoreClient: Firestore;
+
+  private document: DocumentReference<DocumentData> | null;
+
+  private messages: BaseMessage[];
+
+  constructor({
+    collectionName,
+    sessionId,
+    userId,
+    appIdx = 0,
+    config,
+  }: FirestoreDBChatMessageHistory) {
+    super();
+    this.collectionName = collectionName;
+    this.sessionId = sessionId;
+    this.userId = userId;
+    this.document = null;
+    this.appIdx = appIdx;
+    this.messages = [];
```

Comment:
I don't think this should be cached?

Reply:
removed this.messages

Reply:
userId can help admins to define firestore read/write rules

Reply:
for e.g  the following rule ensures only the userId can read messages at    

```
match /chats/{sessionId}/messages/{messageId} {
      allow read: if request.auth.uid == resource.data.createdBy;
     }
     ```